drivers: video: common: fix video_closest_frmival() fie.index

josuah opened the pull request:
Downstream:
- #88003

Fix bug introduced in 46a262ffe62d5e29bf54453de34104779d3e7385 where the fie.index field was expected to be incremented by the driver, while it is the responsibility of the caller to increment it.

Fixes #89663 
Split out of #88003

josuah committed:
drivers: video: common: fix video_closest_frmival() fie.index

Fix bug introduced in 46a262ffe62d5e29bf54453de34104779d3e7385 where the
fie.index field was expected to be incremented by the driver, while it
is the responsibility of the caller to increment it.

Fixes #89663

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented:
Example of implementation that correctly keep `fie.index` unchanged:

https://github.com/zephyrproject-rtos/zephyr/blob/af77efb77e912e109e8a9326c0b522c5058d9218/drivers/video/ov5640.c#L1195-L1221

josuah committed:
drivers: video: common: optimizations for video_closest_frmival()

In video_closest_frmival(), immediately stop searching when an exact
match is found, as a small performance optimization. Variables
that could be computed only once were moved further outside.

Signed-off-by: Josuah Demangeon <me@josuah.net>

avolmat-st commented on drivers/video/video_common.c:
```diff
@@ -153,10 +153,16 @@ void video_closest_frmival(const struct device *dev, enum video_endpoint_id ep,
 		a = video_frmival_nsec(&desired);
```
One more small optimization if I get it right. This first calculation of the desired frmival (a) only has to be done once, so should be done before entering the while loop.

josuah replied:
Applied thank you!

avolmat-st commented on drivers/video/video_common.c:
```diff
@@ -153,10 +153,16 @@ void video_closest_frmival(const struct device *dev, enum video_endpoint_id ep,
 		if (diff_nsec < best_diff_nsec) {
 			best_diff_nsec = diff_nsec;
 			match->index = fie.index;
 			memcpy(&match->discrete, &tmp, sizeof(tmp));
```
what about simply doing match->discrete = tmp here ? Meaning, simply letting the compiler decide what is the best way to do that, and figure out optimization, depending on alignments etc. There is good chances that it would end up to be same as memcpy in this case but this is also I think more easy to read.

josuah replied:
Much better I agree!

josuah committed:
drivers: video: common: turn ASSERT(false) into CODE_UNREACHABLE

The assert message was not very helpful, CODE_UNREACHABLE is more
readable and requires fewer effort while proofreading.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented:
force-push:
- Apply extra improvement suggested
